[jit] Initial use RRef in TorchScript

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33263 [rpc] allow RRef local creation with IValue objects
* **#33190 [jit] Initial use RRef in TorchScript**
* #32992 [jit] Add RRef to IValue and JIT type system
* #33189 [rpc] Switch RRef to be managed by intrusive_ptr

This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

Differential Revision: [D19871244](https://our.internmc.facebook.com/intern/diff/D19871244)

wanchaol committed:
[jit] Initial use RRef in TorchScript

This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Initial use RRef in TorchScript"


This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Initial use RRef in TorchScript"


This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

[ghstack-poisoned]

suo commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -21,6 +21,9 @@
 #include <torch/csrc/utils/auto_gil.h>
 #include <torch/csrc/utils/pybind.h>
 #include <torch/csrc/utils/six.h>
+#ifdef USE_DISTRIBUTED
```
What's the reason for this macro? It seems weird to have the checks here but nowhere else…if we weren't compiling with the distributed package, how would the rest of it work?

wanchaol replied:
Yeah the pickling support for rref is covered by @zhaojuanmao PR https://github.com/pytorch/pytorch/pull/32959. 

For the reason of this macro, I think right now default libtorch is not shipped together with distributed support, so everything for the distributed package get built guarded by this `USE_DISTRIBUTED` macro. cc @pritamdamania @mrshenli for the exact reason why we not default building the distributed package yet. 

If we weren't compiling with the distributed package, we will not include the corresponding header and JIT will just error out to user saying "RRef is only supported with the distributed package"

wanchaol committed:
Update on "[jit] Initial use RRef in TorchScript"


This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

[ghstack-poisoned]

suo reviewed (approved):
Looks reasonable enough to me. In this stack, one element that is missing is a story for how we serialize rrefs. I assume that's covered by @zhaojuanmao's PRs?

wanchaol committed:
Update on "[jit] Initial use RRef in TorchScript"


This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Initial use RRef in TorchScript"


This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

Differential Revision: [D19871244](https://our.internmc.facebook.com/intern/diff/D19871244)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Initial use RRef in TorchScript"


This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

Differential Revision: [D19871244](https://our.internmc.facebook.com/intern/diff/D19871244)

[ghstack-poisoned]

houseroad commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -91,8 +91,9 @@ def check_bc(new_schema_dict):
             if not line:
                 break
-            if "torch.classes" in line:
+            if "torch.classes" or "RRef" in line:
```
this should be 

```
if "torch.classes" in line or "RRef" in line:
```

wanchaol committed:
Update on "[jit] Initial use RRef in TorchScript"


This enable the initial RRef type to be used inside TorchScript, user
could pass a python RRef into a torchscript function and call to_here
inside. Specifically, this PR:

- Add RRef schema type parsing
- Add python interop for RRef in Python and into JIT
- register to_here op in register_distributed_ops

More support for RRef in TorchScript will be added in future PRs

Differential Revision: [D19871244](https://our.internmc.facebook.com/intern/diff/D19871244)

[ghstack-poisoned]
